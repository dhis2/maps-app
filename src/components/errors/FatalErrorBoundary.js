import React, { Component } from 'react';
import PropTypes from 'prop-types';
import i18n from '@dhis2/d2-i18n';
import { withStyles } from '@material-ui/core/styles';
import InfoIcon from '@material-ui/icons/InfoOutlined';

const bgColor = '#F4F6F8',
    secondaryTextColor = '#494949';

const replaceNewlinesWithBreaks = text =>
    text
        .split('\n')
        .reduce((out, line, i) => [...out, line, <br key={i} />], []);

const styles = {
    mask: {
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,

        overflow: 'auto',
        overflowY: 'auto',

        backgroundColor: bgColor,

        display: 'flex',

        minWidth: 640,
        minHeight: 480,

        alignItems: 'center',
        justifyContent: 'center',
    },
    container: {
        textAlign: 'center',
        color: 'black',
    },
    icon: {
        width: 96,
        height: 96,
        color: '#B0BEC5',
        marginBottom: 24,
    },
    message: {
        fontSize: '24px',
        marginBottom: 24,
    },
    link: {
        fontSize: '18px',
        textDecoration: 'underline',
        cursor: 'pointer',
        marginBottom: 24,
    },
    drawerToggle: {
        fontSize: '12px',
        color: secondaryTextColor,
        textDecoration: 'underline',
        cursor: 'pointer',
        marginBottom: 12,
    },
    drawerVisible: {
        padding: 8,
        display: 'block',
        height: 150,
        width: 500,
        overflow: 'auto',
        overflowY: 'auto',
        border: `1px solid ${secondaryTextColor}`,
        textAlign: 'left',
    },
    drawerHidden: {
        display: 'none',
    },
    errorIntro: {
        fontSize: '12px',
        lineHeight: 1.2,
        color: secondaryTextColor,
        marginBottom: 8,
        fontFamily: 'Menlo, Courier, monospace !important',
    },
    errorDetails: {
        fontSize: '12px',
        lineHeight: 1.2,
        color: 'red',
        fontFamily: 'Menlo, Courier, monospace !important',
    },
};

class FatalErrorBoundary extends Component {
    constructor(props) {
        super(props);
        this.state = {
            error: null,
            errorInfo: null,
            drawerOpen: false,
        };
    }

    componentDidCatch(error, info) {
        this.setState({
            error,
            errorInfo: info,
        });
    }

    toggleTechInfoDrawer = () => {
        this.setState({
            drawerOpen: !this.state.drawerOpen,
        });
    };

    render() {
        const { classes, children } = this.props;
        if (this.state.error) {
            return (
                <div className={classes.mask}>
                    <div className={classes.container}>
                        <InfoIcon className={classes.icon} />
                        <div className={classes.message}>
                            {i18n.t('Something went wrong')}
                        </div>
                        <div
                            className={classes.link}
                            onClick={() => window.location.reload()}
                        >
                            {i18n.t('Refresh to try again')}
                        </div>
                        <div
                            className={classes.drawerToggle}
                            onClick={this.toggleTechInfoDrawer}
                        >
                            {this.state.drawerOpen
                                ? i18n.t('Hide technical details')
                                : i18n.t('Show technical details')}
                        </div>
                        <div
                            className={
                                this.state.drawerOpen
                                    ? classes.drawerVisible
                                    : classes.drawerHidden
                            }
                        >
                            <div className={classes.errorIntro}>
                                {i18n.t(
                                    'An error occurred in the DHIS2 {{AppName}} application.',
                                    { AppName: 'Maps' }
                                )}
                                <br />
                                {i18n.t(
                                    'The following information may be requested by technical support.'
                                )}
                            </div>
                            <div className={classes.errorDetails}>
                                {[
                                    replaceNewlinesWithBreaks(
                                        this.state.error.stack +
                                            '\n---' +
                                            this.state.errorInfo.componentStack
                                    ),
                                ]}
                            </div>
                        </div>
                    </div>
                </div>
            );
        }

        return children;
    }
}

FatalErrorBoundary.propTypes = {
    classes: PropTypes.object.isRequired,
    children: PropTypes.node.isRequired,
};

export default withStyles(styles)(FatalErrorBoundary);
