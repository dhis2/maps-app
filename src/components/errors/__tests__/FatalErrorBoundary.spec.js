import React from 'react';
import FatalErrorBoundary from '../FatalErrorBoundary';
import { mount } from 'enzyme';

const Something = () => {
    // Placeholder
    return null;
};

describe('FatalErrorBoundary', () => {
    it('Should render the normal tree when no error occurs', () => {
        const wrapper = mount(
            <FatalErrorBoundary>
                <div>
                    <span id="testme">Hello there!</span>
                </div>
            </FatalErrorBoundary>
        );

        expect(wrapper.find('span#testme').length).toBe(1);
    });

    it('Should render the error mask when an error is thrown', () => {
        const wrapper = mount(
            <FatalErrorBoundary>
                <Something />
            </FatalErrorBoundary>
        );

        expect(wrapper.find(Something).length).toBe(1);

        const testErrorText =
            'This is a test of the emergency alert system.  This is only a test.';
        wrapper.find(Something).simulateError(new Error(testErrorText));
        expect(wrapper.contains('Something went wrong')).toBe(true);
        expect(wrapper).toMatchSnapshot(); // NOTE: This snapshot will change if the .simulateError location changes in this file, since the snapshot includes the callstack!
    });
});
