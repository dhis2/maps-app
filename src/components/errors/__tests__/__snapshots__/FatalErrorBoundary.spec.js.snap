// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatalErrorBoundary Should match the snapshot tree when error is manually invoked 1`] = `
<div
  className="mask"
>
  <div
    className="container"
  >
    <InfoOutlinedIcon
      className="icon"
    />
    <div
      className="message"
    >
      Something went wrong
    </div>
    <div
      className="link"
      onClick={[Function]}
    >
      Refresh to try again
    </div>
    <div
      className="drawerToggle"
      onClick={[Function]}
    >
      Show technical details
    </div>
    <div
      className="drawerHidden"
    >
      <div
        className="errorIntro"
      >
        An error occurred in the DHIS2 Maps application.
        <br />
        The following information may be requested by technical support.
      </div>
      <div
        className="errorDetails"
      >
        This is the stack
        <br
          key="0"
        />
        ---This is the component stack
        <br
          key="1"
        />
      </div>
    </div>
  </div>
</div>
`;
